### PR TITLE
LatLng2Place library using maps API

### PR DESCRIPTION
**Pull request by pradh**

### CODE REVIEW

File: util/latlng2place_mapsapi.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Library that converts a lat/lng into one or more place DCIDs using Maps API.
+
+Currently supports resolution from lat/lng to admin-area2 (county-equivalent)
+places, admin-area1 (state-equivalent) places and countries.
+
+IMPORTANT NOTE: Please do not hardcode the API key in the calling programs,
+instead pass it as arguments while running them.
```
Comment (beets) [inline util/latlng2place_mapsapi.py:20]:
```suggestion
instead pass it as flag arguments while running them.
```

File: util/latlng2place_mapsapi.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                        self._latlng2place[latlng] = parts[2:]
+            except Exception:
+                pass
+            self._cf = open(cache_file, 'a+')
+
+    def _url(self):
+        return 'https://maps.googleapis.com/maps/api/geocode/json?sensor=false&key=' + self._api_key
+
+    def _parse_result(self, result):
+        """Parse resultsf rom Cloud reverse geocoding service."""
```
Comment (beets) [inline util/latlng2place_mapsapi.py:93]:
```suggestion
        """Parse results from Cloud reverse geocoding service."""
```

File: util/latlng2place_mapsapi.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    return ''
+
+
+class Resolver(object):
+    """Resolves lat lngs to DCIDs."""
+
+    def __init__(self, api_key, cache_file='', cache_only=False):
+        assert api_key, 'Resolver() needs a valid API key'
+        self._api_key = api_key
+        self._cache_file = cache_file
```
Comment (beets) [inline util/latlng2place_mapsapi.py:73]: this doesn't seem necessary since it's only used in the constructor

File: util/latlng2place_mapsapi.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        self._cache_only = cache_only
+        self._latlng2place = {}
```
Comment (beets) [inline util/latlng2place_mapsapi.py:75]: nit: please document these class variables, and declare _cf here too for readability.

File: util/latlng2place_mapsapi.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+"""Library that converts a lat/lng into one or more place DCIDs using Maps API.
+
+Currently supports resolution from lat/lng to admin-area2 (county-equivalent)
+places, admin-area1 (state-equivalent) places and countries.
+
+IMPORTANT NOTE: Please do not hardcode the API key in the calling programs,
+instead pass it as arguments while running them.
+
+Example:
+  ll2p = latlng2place_mapsapi.Resolver(api_key='XYZ')
```
Comment (beets) [inline util/latlng2place_mapsapi.py:23]: also mention how to use caching?

File: util/latlng2place_mapsapi.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  try:
+    dcids = ll2p.resolve(12.0, -34.0)
+  except:
+    logging.error('Failed to resolve 12.0, -34.0')
+  ... more calls to llp.resolve() ...
```
Comment (beets) [inline util/latlng2place_mapsapi.py:28]:
```suggestion
  ... more calls to ll2p.resolve() ...
```

File: util/latlng2place_mapsapi.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+# Allows the following module imports to work when running as a script
+sys.path.append(
+    os.path.dirname(os.path.dirname(os.path.dirname(
+        os.path.abspath(__file__)))))
+
+import aa_isocode2dcid
+import alpha2_to_dcid
+
+_LOCK = threading.Lock()
+_PLACE_TYPES = ['administrative_area2', 'administrative_area1', 'country']
```
Comment (beets) [inline util/latlng2place_mapsapi.py:45]: please document.

File: util/latlng2place_mapsapi.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+        # Extract short-names.
+        short_names = {}
+        for place in result['address_components']:
+            if 'administrative_area_level_2' in place['types']:
+                short_names['administrative_area2'] = place['short_name']
+            elif 'administrative_area_level_1' in place['types']:
+                short_names['administrative_area1'] = place['short_name']
+            elif 'country' in place['types']:
+                short_names['country'] = place['short_name']
```
Comment (beets) [inline util/latlng2place_mapsapi.py:114]: i wonder if we could do something like:

```
for place in result['address_components']:
  for pt in _PLACE_TYPES:
    if pt in place['types']
      short_names[pt] = place['short_name']
      break
```
noting the ordering constraint in `_PLACE_TYPES`

**Review by beets — APPROVED** [review]
cool library!

File: util/latlng2place_mapsapi.py (hunk 3 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py]: It was used in line 169 before, but replaced it with _cf

File: util/latlng2place_mapsapi.py (hunk 7 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py:52]: done

File: util/latlng2place_mapsapi.py (hunk 6 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py]: Done

File: util/latlng2place_mapsapi.py (hunk 5 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py:23]: Done

File: util/latlng2place_mapsapi.py (hunk 1 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py]: Done

File: util/latlng2place_mapsapi.py (hunk 4 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py:91]: Done

File: util/latlng2place_mapsapi.py (hunk 2 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py]: Done

File: util/latlng2place_mapsapi.py (hunk 8 above)
Comment (pradh) [reply util/latlng2place_mapsapi.py]: Good idea!  (I'd largely blindly ported http://google3/datacommons/mcf/lib/latlng2place.py;l=64-103;rcl=481245757)

**Review by pradh — COMMENTED** [review]
Thanks for the review @beets !

**Comment by pradh** [comment]
I will mark comments resolved after approval.

**Comment by pradh** [comment]
Oh, there's approval - marking things as resolved.

**Pull request merged**